src/backend/distributed/commands/dependencies.c

```diff
@@ -0,0 +1,196 @@
+ *
+ * This is solved by creating the dependencies in an idempotent manner, either via
+ * postgres native CREATE IF NOT EXISTS, or citus helper functions.
+ */
+void
+EnsureDependenciesExistsOnAllNodes(const ObjectAddress *target)
+{
+	const uint64 connectionFlag = FORCE_NEW_CONNECTION;
+
+	/* local variables to work with dependencies */
+	List *dependencies = NIL;
+	ListCell *dependencyCell = NULL;
+
+	/* local variables to collect ddl commands */
+	List *ddlCommands = NULL;
+
+	/* local variables to work with worker nodes */
+	List *workerNodeList = NULL;
+	ListCell *workerNodeCell = NULL;
+	List *connections = NULL;
+	ListCell *connectionCell = NULL;
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	dependencies = GetDependenciesForObject(target);
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		ddlCommands = list_concat(ddlCommands,
+								  GetDependencyCreateDDLCommands(dependency));
+	}
+	if (list_length(ddlCommands) <= 0)
+	{
+		/* no ddl commands to be executed */
+		return;
+	}
+
+	/*
+	 * collect and connect to all applicable nodes
+	 */
+	workerNodeList = ActivePrimaryNodeList();
+	if (list_length(workerNodeList) <= 0)
+	{
+		/* no nodes to execute on */
+		return;
+	}
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		MultiConnection *connection = NULL;
+
+		char *nodeName = workerNode->workerName;
+		uint32 nodePort = workerNode->workerPort;
+
+		connection = GetNodeUserDatabaseConnection(connectionFlag, nodeName,
+												   nodePort,
+												   CitusExtensionOwnerName(),
+												   NULL);
+
+		connections = lappend(connections, connection);
+	}
+
+	/*
+	 * create dependency on all nodes
+	 */
+	foreach(connectionCell, connections)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		ExecuteCriticalRemoteCommandList(connection, ddlCommands);
+	}
+
+	/*
+	 * mark all objects as distributed
+	 */
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		markObjectDistributed(dependency);
+	}
+
+	/*
+	 * disconnect from nodes
+	 */
+	foreach(connectionCell, connections)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		CloseConnection(connection);
+	}
+}
+
+
+/*
+ * GetDependencyCreateDDLCommands returns a list (potentially empty or NIL) of ddl
+ * commands to execute on a worker to create the object.
+ */
+static List *
+GetDependencyCreateDDLCommands(const ObjectAddress *dependency)
+{
+	switch (getObjectClass(dependency))
+	{
+		case OCLASS_SCHEMA:
+		{
+			const char *schemaDDLCommand = CreateSchemaDDLCommand(dependency->objectId);
+
+			if (schemaDDLCommand == NULL)
+			{
+				/* no schema to create */
+				return NIL;
+			}
+
+			return list_make1((void *) schemaDDLCommand);
+		}
+
+		default:
+		{
+			return NIL;
```

Comment:
Do we ever expect to get `NIL`? Given that we're guarding the objects via `SupportedDependencyByCitus ()`, I cannot see a case where this function would return `NIL`. If I'm not missing anything, it'd be nice to add a comment or maybe an assert as well.

Reply:
this is primarily a safety check, which might be actually too safe.

Guessing an error here will make it directly clear for the developer the guard of `SupportedDependencyByCitus`.

There will be some dependencies 'supported' that will result in a `NIL` returned from this function though. Namely
 - `public` schema
 - array types
 - composite relations

When `NIL` is returned the object in question will not be marked distributed in `pg_dist_object` to keep that for actual object we have created and need to create on new nodes in the future. The potential of returning `NIL` is already documented on top.